File: clients/javascript/test-reader.js
/**
 * Universal Crash Analytics API - Crash Reader Test
 * Test the crash reading functionality
 */

require('dotenv').config();
const CrashReader = require('./crash_reader');

async function testCrashReader() {
  console.log('🧪 Testing Crash Reader Functionality');
  console.log('=====================================');
  
  // Initialize the crash reader
  const reader = new CrashReader({
    appName: 'test-app',
    appVersion: 'v1.0.0'
  });
  
  try {
    // Test 1: Read recent crashes
    console.log('\n📊 Test 1: Reading recent crashes (last 24 hours)');
    const recentCrashes = await reader.getRecentCrashes(24);
    console.log(`✅ Found ${recentCrashes.length} recent crashes`);
    
    if (recentCrashes.length > 0) {
      console.log('📋 Sample crash report:');
      console.log(JSON.stringify(recentCrashes[0], null, 2));
    }
    
    // Test 2: Get crash statistics
    console.log('\n📈 Test 2: Getting crash statistics (last 30 days)');
    const stats = await reader.getCrashStats(30);
    console.log('✅ Crash statistics retrieved successfully');
    console.log(`📊 Total crashes: ${stats.total_crashes}`);
    console.log(`👥 Unique users affected: ${stats.unique_users}`);
    console.log(`🖥️  Platforms: ${JSON.stringify(stats.platforms)}`);
    console.log(`📱 Versions: ${JSON.stringify(stats.versions)}`);
    console.log(`⚠️  Top errors: ${Object.keys(stats.top_errors).slice(0, 3)}`);
    console.log(`⏰ Time distribution: ${JSON.stringify(stats.time_distribution)}`);
    
    // Test 3: Search for specific errors
    console.log('\n🔍 Test 3: Searching for specific errors');
    const errorCrashes = await reader.getCrashesByError('Test crash', 7);
    console.log(`✅ Found ${errorCrashes.length} crashes with 'Test crash' in the last 7 days`);
    
    // Test 4: Read with pagination
    console.log('\n📄 Test 4: Testing pagination');
    const paginatedReports = await reader.readCrashReports({
      limit: 5,
      offset: 0,
      days: 30
    });
    console.log(`✅ Retrieved ${paginatedReports.data.length} reports`);
    console.log(`📊 Pagination info: ${JSON.stringify(paginatedReports.pagination)}`);
    
    // Test 5: Filter by version
    console.log('\n🔧 Test 5: Filtering by version');
    const versionReports = await reader.readCrashReports({
      version: 'v1.0.0',
      days: 30
    });
    console.log(`✅ Found ${versionReports.data.length} crashes for version v1.0.0`);
    
    console.log('\n🎉 All crash reader tests completed successfully!');
    console.log('📚 Your crash reading functionality is working correctly');
    
  } catch (error) {
    console.error('❌ Crash reader test failed:', error.message);
    process.exit(1);
  }
}

// Run the test
testCrashReader();

File: clients/python/test_reader.py
#!/usr/bin/env python3
"""
Universal Crash Analytics API - Crash Reader Test
Test the crash reading functionality
"""

import sys
import os
from crash_reader import CrashReader

def test_crash_reader():
    """Test the crash reader functionality"""
    print("🧪 Testing Crash Reader Functionality")
    print("=====================================")
    
    # Initialize the crash reader
    reader = CrashReader({
        'appName': 'test-python-app',
        'appVersion': 'v1.0.0'
    })
    
    try:
        # Test 1: Read recent crashes
        print("\n📊 Test 1: Reading recent crashes (last 24 hours)")
        recent_crashes = reader.get_recent_crashes(24)
        print(f"✅ Found {len(recent_crashes)} recent crashes")
        
        if recent_crashes:
            print("📋 Sample crash report:")
            import json
            print(json.dumps(recent_crashes[0], indent=2))
        
        # Test 2: Get crash statistics
        print("\n📈 Test 2: Getting crash statistics (last 30 days)")
        stats = reader.get_crash_stats(30)
        print("✅ Crash statistics retrieved successfully")
        print(f"📊 Total crashes: {stats['total_crashes']}")
        print(f"👥 Unique users affected: {stats['unique_users']}")
        print(f"🖥️  Platforms: {stats['platforms']}")
        print(f"📱 Versions: {stats['versions']}")
        print(f"⚠️  Top errors: {list(stats['top_errors'].keys())[:3]}")
        print(f"⏰ Time distribution: {stats['time_distribution']}")
        
        # Test 3: Search for specific errors
        print("\n🔍 Test 3: Searching for specific errors")
        error_crashes = reader.get_crashes_by_error("Test crash", 7)
        print(f"✅ Found {len(error_crashes)} crashes with 'Test crash' in the last 7 days")
        
        # Test 4: Read with pagination
        print("\n📄 Test 4: Testing pagination")
        paginated_reports = reader.read_crash_reports({
            'limit': 5,
            'offset': 0,
            'days': 30
        })
        print(f"✅ Retrieved {len(paginated_reports['data'])} reports")
        print(f"📊 Pagination info: {paginated_reports['pagination']}")
        
        # Test 5: Filter by version
        print("\n🔧 Test 5: Filtering by version")
        version_reports = reader.read_crash_reports({
            'version': 'v1.0.0',
            'days': 30
        })
        print(f"✅ Found {len(version_reports['data'])} crashes for version v1.0.0")
        
        print("\n🎉 All crash reader tests completed successfully!")
        print("📚 Your crash reading functionality is working correctly")
        
    except Exception as e:
        print(f"❌ Crash reader test failed: {e}")
        sys.exit(1)

if __name__ == '__main__':
    test_crash_reader()
